release-23.1: improve tenant server start/stop in shared-process multitenancy

knz opened the pull request:
Backports:

- 9/9 commits from #100436.
- 1/1 commit from #99958.
- 11/11 commits from #101296.
- 1/1 commit from #101459.

----

Release justification: prevents SQL app disruption during node restarts
Epic: CRDB-23559

knz committed:
server: gracefully shut down secondary tenant servers

This change ensures that tenant servers managed by the server
controller receive a graceful drain request as part of the graceful
drain process of the surrounding KV node.

This change, in turn, ensures that SQL clients connected to these
secondary tenant servers benefit from the same guarantees (and
graceful periods) as clients to the system tenant.

Release note: None

knz committed:
server: unexport some functions

Release note: None

knz committed:
server: save a reference to BaseConfig in SQLServerWrapper

Release note: None

knz committed:
server: extend the `onDemandServer` interface

ahead of splitting "new" vs "start" during construction.

Release note: None

knz committed:
server: lift `reportTenantInfo` into `*SQLServerWrapper`

Release note: None

knz committed:
server: separate "new" from "start" for tenant servers

This peels the call to "start" from the `newTenantServer`
interface and pulls it into the orchestration retry loop.

This change also incidentally reveals an earlier misdesign: we are
calling `newTenantServer` _then_ `start` in the same retry loop. If
`new` succeeds but `start` fails, the next retry will call
`newTenantServer` again *with the same stopper*, which will leak
closers from the previous call to `new`.

Release note: None

knz committed:
server: fix the tenant server error handling

Prior to this patch, if an error occurred during the initialization or
startup of a secondary tenant server, the initialization would leak
state into the stopper defined for that tenant. Generally, reusing
a stopper across server startup failures is not safe (and API
violation).

This patch fixes it by decoupling the intermediate stopper used for
orchestration from the one used per tenant server.

Release note: None

knz committed:
serverccl: simplify TestServerStartupGuardrails

Prior to this patch, the test was not cleaning up its server stopper
reliably at the end of each sub-test. This patch fixes it.

Release note: None

knz committed:
serverccl: make TestServerControllerMultiNodeTenantStartup faster

Release note: None

knz committed:
serverccl: clarify the progress inside TestServerControllerMultiNodeTenantStartup

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/101089)
<!-- Reviewable:end -->

knz commented:
I plan to only mark this ready once we have a fix for #100989.

knz committed:
server: lift the orchestration code under an interface

This commit extracts the core logic from
`server_controller_orchestration.go` into an
interface (`serverOrchestrator`) that can be mocked. We will use this
in testing.

This extraction is also useful because it exposes a few shortcomings
in the current implementation (wrt server shutdown). This makes it
easier to fix them in a later commit.

Release note: None

knz committed:
server: move the main tenant orchestration logic to its own file

This moves `(o *channelOrchestrator) startControlledServer` alongside
the other methods of `channelOrchestrator`.
There is further no code change.

Release note: None

knz committed:
server: split graceful vs ungraceful stop for tenant servers

Prior to this patch, immediate shutdown requests for secondary tenant
servers were serviced using the graceful path. This was incorrect, and
resulted in excessively long shutdown sequences in tests in some
cases.

This patch fixes it.

Release note: None

knz committed:
server: extra logging for server controller

Release note: None

knz committed:
server: avoid naked writes to a channel

The sever controller was writing to
`useGracefulDrainDuringTenantShutdown` unconditionally. As a result,
if two or more expedited stops caught up with a graceful stop, some of
the writes to the channel could be blocked. This would never happen on
a running system, but could be exercised in tests.

This patch fixes this by making the write unblocking. This generally
makes the code more readable anyway, since naked channel writes are
bound to raise eyebrows anyways.

Release note: None

knz committed:
server: add some shutdown logging

This helps while troubleshooting tests.

Release note: None

knz committed:
server: fix a possible deadlock during tenant server shutdown

Prior to this patch, the server orchestrator was unable to notice when
the surrounding server was stopping after it had started a graceful
drain already. This is because once the `propagate-close` task (and
its `select`) monitoring various stop signals noticed a request for
graceful drains, it would stop monitoring the other stop signals to
focus exclusively on the graceful drain; thereby missing when the
stopper for the surrounding server was quiescing.

The fix is to run the monitor for graceful and ungraceful shutdowns in
separate tasks, which this patch does.

This goes 80% of the way towards de-flaking `TestServerStartStop` in
`pkg/ccl/serverccl`, which was exhibiting the deadlock described
above. With this fix in place, the deadlock disappears entirely.
However, this reveals _another_ bug in the SQL layer which also needs
to be fixed to consider `TestServerStartStop` stable. We will do this
in a separate commit.

Release note: None

knz committed:
serverccl: add additional logging to `TestServerStartStop`

This also ensures the test fails quickly if there is a deadlock while
the server is shutting down.

(This makes the timeout for this test shorter than the standard
timeout for the stopper.Stop method, which is 15 minutes.)

Release note: None

knz committed:
util/mon: better logging

Release note: None

knz committed:
sqlstats: paper over a memory monitor bug

When a hard stop (e.g. test exists) catches up with an ongoing
graceful drain, some bytes do not get released properly in a memory
monitor. This triggers a panic during shutdown, since the byte monitor
code verifies that all allocated bytes have been released.

This bug is relatively hard to trigger because in most cases,
a server is shut down either only using a graceful drain, or only
using a hard stop, but not both.

`TestServerStartStop` happens to do this and this is where this
problem was caught. We are now tracking it as issue #101297.

Until that issue is fixed, this commit papers over the problem by
removing the assertion in the byte monitor.

Release note: None

knz committed:
server: avoid a race condition in tenant server orchestration

Prior to this patch, the task that was responsible for propagating a
graceful drain captured a variable supposed to reference the tenant
server, but before this variable was assigned.

This created a possible race condition, in the unlikely case case
where the server startup would fail _and_ a graceful drain would be
requested, concurrently.

This patch fixes it by only starting to propagate graceful drains
after the server is fully initialized. (But before it starts accepting
clients, so that we don't create a window of time where clients can
connects but graceful drains don't propagate.)

This is also achieved by extracting the two shutdown tasks into
separate functions, to clarify the flow of parameters.

Release note: None

knz committed:
server: deflake TestClusterVersionUpgrade under race

Release note: None

jeffswenson reviewed (approved):
LGTM
